File: ts/src/134.ts
//115. Distinct Subsequences
function numDistinct(s: string, t: string): number {
  let m = s.length;
  let n = t.length;
  const MOD = 1000000007n; // Use bigint

  const dp: bigint[][] = Array.from({ length: m + 1 }, () =>
    Array(n + 1).fill(0n),
  );
  for (let i = 0; i <= m; i++) dp[i][0] = 1n;

  for (let i = 1; i <= m; ++i) {
    for (let j = 1; j <= n; ++j) {
      if (s[i - 1] == t[j - 1])
        dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j]) % MOD;
      else dp[i][j] = dp[i - 1][j];
    }
  }
  return Number(dp[m][n]);
}

//sub: https://leetcode.com/problems/distinct-subsequences/submissions/1557735139/
//sol: https://leetcode.com/problems/distinct-subsequences/solutions/6476059/115-distinct-subsequences-by-nalindalal2-8ksp/

File: rust/src/134.rs
//134. Gas Station
impl Solution {
    pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
        let n = gas.len();
        let mut total_tank = 0;
        let mut current_tank = 0;
        let mut start = 0;

        for i in 0..n {
            let diff = gas[i] - cost[i];
            total_tank += diff;
            current_tank += diff;

            if current_tank < 0 {
                // Can't reach the next station
                start = (i + 1) as i32; // Start from the next station
                current_tank = 0; // Reset the current tank
            }
        }

        if total_tank >= 0 {
            start
        } else {
            -1
        } // Corrected condition
    }
}

//sub: https://leetcode.com/problems/gas-station/submissions/1560988864/
//sol: https://leetcode.com/problems/gas-station/solutions/6488419/134-gas-station-by-nalindalal2004-n2ug/
